### PR TITLE
Move menus into a separate activity

### PR DESCRIPTION
**Pull request by lidavidm**
This is a work-in-progress; best not to merge until we can clean up code style/etc.

TODO
- [x] Better styling
- [ ] Implement that slick transition
  - [x] Step 1: Use ToolBar, not ActionBar
- [x] Figure out why app crashes when you rotate the screen when viewing the menu
- [x] Figure out why activity title disappears after rotating it twice
- [x] Color the navbar/statusbar as well
- [x] Create a "no menu available" layout

IDEAS
- [ ] Show the menu for the upcoming meal (i.e. don't show breakfast at this time; show a card at the bottom for tomorrow's breakfast)

device-2015-03-24-204416
device-2015-03-24-204432

### CODE REVIEW

File: app/build.gradle [hunk 1]
```diff
@@ -20,6 +20,8 @@ android {
 }
 
 dependencies {
-    compile fileTree(dir: 'libs', include: ['*.jar'])
```
Comment (lidavidm) [inline app/build.gradle:23]: Not sure what the purpose of this line was?

File: app/src/main/java/is/genki/bigredapp/android/DiningListFragment.java [hunk 2]
```diff
@@ -30,7 +31,7 @@
 public class DiningListFragment extends Fragment {
 
     private static final String BASE_URL = "http://redapi-tious.rhcloud.com/dining";
-    private static final String[] MEALS_LIST = {"Breakfast", "Lunch", "Dinner"};
+    public static final String[] MEALS_LIST = {"Breakfast", "Lunch", "Dinner"};
```
Comment (lidavidm) [inline app/src/main/java/is/genki/bigredapp/android/DiningListFragment.java]: So brunch isn't handled, right?

File: app/src/main/java/is/genki/bigredapp/android/GetRequest.java [hunk 3]
```diff
@@ -80,4 +82,14 @@ private String downloadUrl(String myUrl) throws IOException {
         }
     }
 
+    public static boolean isConnected(Activity activity) {
```
Comment (lidavidm) [inline app/src/main/java/is/genki/bigredapp/android/GetRequest.java:90]: I factored this out here, but I haven't changed the original method to use this one yet. (I wish Java had C#'s extension methods.) Regardless, if we use an external library for this sort of thing it won't matter anymore.

File: app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+        }
+
+        Bundle extras = getIntent().getExtras();
+        if (extras != null) {
+            final String diningHall = extras.getString(KEY_DINING_HALL);
+            String url = extras.getString(KEY_DINING_HALL_URL);
+
+            setTitle(diningHall);
+
+            if (GetRequest.isConnected(this)) {
```
Comment (lidavidm) [inline app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java]: I don't know if it matters that we wait until the activity is created to fire off the HTTP request.

File: app/build.gradle (hunk 1 above)
Comment (ghost) [inline app/build.gradle:23]: Looks like this was added automatically by AS on project init. Seems it's just supposed to provide out of the box added-library-jar compiling, as you probably guessed and thus wondered why it was necessary. It's not, so good call :D

File: app/src/main/java/is/genki/bigredapp/android/DiningListFragment.java (hunk 2 above)
Comment (ghost) [inline app/src/main/java/is/genki/bigredapp/android/DiningListFragment.java]: Oh good point. I found the source for RedAPI @ https://github.com/mrkev/Iroh and https://github.com/mrkev/redapi. Looking at the repos... maybe it wouldn't be a bad idea to set up a new repository, consulting/working with Kevin, to provide the backend for this app, that could be then used for other platforms too. Just tossing the idea out there.

File: app/src/main/java/is/genki/bigredapp/android/GetRequest.java (hunk 3 above)
Comment (ghost) [inline app/src/main/java/is/genki/bigredapp/android/GetRequest.java:90]: Agreed on the factoring out. It makes sense as a static function.

File: app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java (hunk 4 above)
Comment (ghost) [inline app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java]: I like it :+1:

**Comment by lidavidm** [comment]
Cornell has official color guidelines: https://brand.cornell.edu/color.php

**Comment by lidavidm** [comment]
device-2015-03-25-091855

Changes:
- Color status bar
- Show loading spinner
- Refactoring
- Fix crash when you rotate activity
- Add screen for when no menus are available
- Use different red color

**Comment by ghost** [comment]
:+1: Awesome ^_^ Will get to pulling this down and going through it later tonight / as soon as I can make time.

**Comment by lidavidm** [comment]
So I have no idea how to get that slick transition to actually work while looking nice. But I've tried. I'll push what I have, and we can revert it if we can't get it to work. What I'm thinking is that the child activity just shouldn't use a toolbar - we should fake our own title. The only app I can think of that implements anything similar is the Google Calendar app.

http://gfycat.com/HotClumsyIslandwhistler

**Comment by lidavidm** [comment]
I gave up. It uses a scale animation now.

Gfycat: http://gfycat.com/CommonSoulfulIcterinewarbler

**Comment by ghost** [comment]
Re: the animations. I really wish I could play around with this stuff and see what I can do, but as you know Motorola refuses to give me 5.0. Yeesh.
Re: menu of breakfast for tomorrow. Just double checked, RedAPI only supports getting the menu for a particular day (I assume it continues to get the day's menu until midnight, or something). I feel bad pressuring Kevin cause we kinda just sprung this project on him. So maybe it would be good to discuss getting our own version set up.
Anyway, I have a PR going at [your fork](https://github.com/lidavidm/bigredapp-android/pull/1) for some random minor stuff.

**Comment by lidavidm** [comment]
The scale animation here works for you right?

**Comment by ghost** [comment]
One sec, I'll post a gif. It looks more like a minimal version of yours.
Also, by the way, not sure why (something to do with me being on API 19 I'm sure), but my action bar has black text instead of white text, in both the DiningListFragment and the DiningListActivity: [picture](https://dl.pushbulletusercontent.com/rimxU2RLRJ3455gWRyERdFFc00za9M6p/Screenshot_2015-03-27-13-46-55.png). Any ideas? I'll do some googling in a moment.

**Comment by lidavidm** [comment]
Yeah, I switched to the Toolbar instead of ActionBar because I thought it
would help with the transition. Apparently not. I think I can still get it
to work, but it'll take more work. Anyways, I haven't themed the toolbar
yet, so if we stick with it we gotta fix that.

I recorded the gif with 5x slower animations, that's why.

**Comment by ghost** [comment]
Ah okay. So, would it be better to just switch back to ActionBar then? I'm not well read on that stuff yet.

As for the animation, right ha I knew it was slowed, but ah didn't know it was that much. Anyway, looking more closely, the things that come for free in 5.0 (e.g. tap on row expanding to fill row, leading into animation) made it seem quite different at first, but yeah, it really is the same animation.

**Comment by lidavidm** [comment]
Yeah, I can revert that commit if we're going to give up on the shared
element transition.

**Comment by ghost** [comment]
Hmm, so the animation would be [this one](https://developer.android.com/design/material/videos/ContactsAnim.mp4), where you would tap on the title and it would expand in size as it went to the Toolbar title?

**Comment by lidavidm** [comment]
That's what I want to achieve... No idea how to do it though. I think we
have to have a separate TextView, use that as the shared element, and have
it overlap with the toolbar as you scroll up or something, like how Play
Music does it. We can get that done later though.

**Comment by ghost** [comment]
Yeah okay. It definitely looks very nice ha. According to [this](http://stackoverflow.com/questions/26567377/are-activity-fragment-transitions-compatible-with-pre-lollipop-devices) it looks like I won't be able to see the effect on my device? But yeah, anyway, down to check out it out later too.
edit: reading more carefully, actually seems it is possible with some custom animations for pre-lollipop, but yeah, that can come later.

**Comment by lidavidm** [comment]
Yeah, and that's complicated xD I think for now using the scale animation is fine. Not many will be using Lollipop anyways.

**Comment by ghost** [comment]
Haha yeah, agreed.

**Comment by lidavidm** [comment]
Alright, reverted to ActionBar instead of Toolbar. Any other changes you'd like?

**Comment by lidavidm** [comment]
http://stackoverflow.com/questions/19003671/how-to-get-a-reverse-animation-of-activityoptions-makescaleupanimation-when-fini suggests that Google Keep animates a fragment or view (presumably the fragment contains a Toolbar) instead of making the card a separate activity. Wouldn't be too bad to implement here, especially seeing as the logic of the new activity is mostly in the fragment. But that can wait.

**Comment by lidavidm** [comment]
See: http://developer.android.com/training/animation/zoom.html

File: app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+                mDiningHallUrl = url;
+
+                if (GetRequest.isConnected(this)) {
+                    // Async Task to get the menu for a dining hall
+                    new GetRequest() {
+                        @Override
+                        protected void onPostExecute(String result) {
+                            try {
+                                List<MealMenu> menus = new ArrayList<>();
+                                if (result == null) throw new JSONException("No result");
```
Comment (ghost) [inline app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java]: I don't think this check is necessary right? The `if (result == null)` line (line 63). It is my understanding that JSONObject construction will throw an exception on a null parameter.

**Comment by ghost** [comment]
Re: animations. Ah yeah that is cool, makes sense what they're doing there for Google Keep. Definitely yeah, we should keep that in mind.
Re: ready PR. Besides the above line note, the only other things I'd like to bring up I have in my PR are here: https://github.com/lidavidm/bigredapp-android/pull/1
I'm totally open for other workflows, but I think doing this ^ style might work out nicely. Could I bother you to go through them?

File: app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java (hunk 5 above)
Comment (lidavidm) [inline app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java]: Oh really? Hmm. I think I saw a case where an exception was thrown, but it wasn't a JSONException. Oh well, we can change it.

**Comment by lidavidm** [comment]
Sure thing, just noticed the PR xD

**Comment by ghost** [comment]
Haha kk thanks ^_^

**Comment by lidavidm** [comment]
Alright, removed the null check and merged your changes.

**Comment by ghost** [comment]
Awesome :D

File: app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java (hunk 5 above)
Comment (ghost) [inline app/src/main/java/is/genki/bigredapp/android/DiningLocationActivity.java]: You're totally right. That case I found was when the request timed out (e.g. for Jansen's Dining Room request, which is currently broken). Sorry about that! Reverted in 8a99992faf339fa5c04b517c21d47566e280a089

**Pull request merged**